File: ios/MenCare/Core/DesignSystem/Component/FormFields.swift
import SwiftUI

/// Una sola famiglia di campi per tutta l'app.
///
/// Prima lo stesso concetto aveva quattro comportamenti diversi (il telefono era
/// validato in registrazione, grezzo nell'editor operatore e inesistente nelle
/// due schede "nuovo cliente"). Qui la regola arriva sempre da
/// `Core/Common/Validation.swift`, che è la copia delle regole del backend:
/// stesso campo = stessa regola = stesso messaggio, ovunque.

// MARK: - Riga messaggio

/// Riga sotto un campo: errore in rosso, altrimenti l'eventuale aiuto in grigio.
/// È l'unico posto in cui si disegna il feedback di un campo.
struct FieldMessageRow: View {
    var error: String?
    var helper: String?

    var body: some View {
        if let error {
            Text(error)
                .font(Typo.bodySmall)
                .foregroundStyle(Color.errorRed)
                .fixedSize(horizontal: false, vertical: true)
        } else if let helper {
            Text(helper)
                .font(Typo.bodySmall)
                .foregroundStyle(Color.textMuted)
                .fixedSize(horizontal: false, vertical: true)
        }
    }
}

// MARK: - Nome

/// Nome o cognome: maiuscola iniziale automatica e la regola dei nomi propri.
struct NameField: View {
    let label: String
    @Binding var text: String
    var placeholder: String?
    var error: String?
    var height: CGFloat = 52
    var outlined: Bool = false

    var body: some View {
        FilledTextField(
            label: label,
            text: $text,
            placeholder: placeholder,
            autocapitalization: .words,
            error: error,
            height: height,
            outlined: outlined
        )
    }
}

// MARK: - Email

/// Tastiera email e nessuna maiuscola automatica: "Mario@…" e "mario@…" sono
/// lo stesso account, e la normalizzazione a minuscolo la fa il view model.
struct EmailField: View {
    let label: String
    @Binding var text: String
    var placeholder: String?
    var leadingSystemImage: String?
    var error: String?
    var height: CGFloat = 52
    var outlined: Bool = false

    var body: some View {
        FilledTextField(
            label: label,
            text: $text,
            placeholder: placeholder,
            leadingSystemImage: leadingSystemImage,
            keyboard: .emailAddress,
            autocapitalization: .never,
            error: error,
            height: height,
            outlined: outlined
        )
    }
}

// MARK: - Telefono

/// Telefono: tastiera numerica, prefisso `+39` mostrato finché il numero non ne
/// porta uno suo, e normalizzazione in E.164 quando il campo perde il fuoco —
/// così quello che si salva è sempre nello stesso formato, da qualsiasi
/// schermata arrivi.
struct PhoneField: View {
    let label: String
    @Binding var text: String
    var placeholder: String?
    var error: String?
    var height: CGFloat = 52
    var outlined: Bool = false

    @FocusState private var focused: Bool

    /// Il prefisso è solo decorazione: sparisce appena il numero ne ha uno
    /// proprio (`+39…`, `0039…`), altrimenti si leggerebbe "+39 +39 347…".
    private var showsPrefix: Bool {
        let trimmed = text.trimmingCharacters(in: .whitespaces)
        return !trimmed.hasPrefix("+") && !trimmed.hasPrefix("00")
    }

    var body: some View {
        FilledTextField(
            label: label,
            text: $text,
            placeholder: placeholder,
            prefix: showsPrefix ? defaultPhoneCountryCode : nil,
            keyboard: .phonePad,
            autocapitalization: .never,
            error: error,
            height: height,
            outlined: outlined,
            focus: $focused
        )
        .onChange(of: focused) { _, isFocused in
            guard !isFocused else { return }
            if let e164 = normalizePhone(text) { text = e164 }
        }
    }
}

// MARK: - Password

/// Password: mostra/nascondi già nel campo, più il misuratore di forza quando
/// serve. Il punteggio arriva da `passwordStrength`, lo stesso del backend:
/// prima registrazione e cambio password lo calcolavano ognuna per conto suo.
struct PasswordField: View {
    let label: String
    @Binding var text: String
    var error: String?
    var showStrength: Bool = false
    var height: CGFloat = 52

    private var strength: PasswordStrength? {
        guard showStrength, !text.isEmpty else { return nil }
        return passwordStrength(text)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            BorderedPasswordField(label: label, text: $text, error: error, height: height)
            if let strength {
                HStack(spacing: 7) {
                    SegmentedMeter(filled: strength.meterFill, total: 4)
                    Text(strength.label)
                        .font(Typo.jost(11))
                        .foregroundStyle(Color.ink)
                }
            }
        }
    }
}

// MARK: - Prezzo

/// Prezzo in euro: tastiera decimale, un solo separatore ammesso e conversione
/// in centesimi con arrotondamento (mai troncamento) fatta dal view model con
/// `parsePriceToCents`. Il filtro a ogni tasto impedisce che "1.2.3" arrivi al
/// parser e finisca salvato come 0,00 €.
struct PriceField: View {
    let label: String
    @Binding var text: String
    var error: String?
    var height: CGFloat = 56

    var body: some View {
        FilledTextField(
            label: label,
            text: Binding(
                get: { text },
                set: { text = sanitizePriceInput($0) }
            ),
            placeholder: "0",
            keyboard: .decimalPad,
            autocapitalization: .never,
            error: error,
            height: height
        )
    }
}

// MARK: - Durata

/// Durata a passi di 5 minuti, con i limiti del backend (da 5 minuti a 8 ore).
struct DurationField: View {
    let label: String
    @Binding var minutes: Int
    var error: String?

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            if !label.isEmpty {
                BrandSectionLabel(text: label)
            }
            HStack {
                Text(L("svc_minutes", minutes))
                    .font(Typo.bodyLarge)
                    .foregroundStyle(Color.ink)
                Spacer()
                stepperButton("minus") {
                    minutes = clampDuration(minutes - ValidationLimits.durationStepMinutes)
                }
                stepperButton("plus") {
                    minutes = clampDuration(minutes + ValidationLimits.durationStepMinutes)
                }
            }
            .padding(.horizontal, 14)
            .frame(height: 56)
            .background(RoundedRectangle(cornerRadius: 14).fill(Color.stone))
            FieldMessageRow(error: error, helper: nil)
        }
    }

    private func stepperButton(_ systemImage: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Circle()
                .fill(Color.bone)
                .frame(width: 28, height: 28)
                .overlay(
                    Image(systemName: systemImage)
                        .font(.system(size: 12, weight: .medium))
                        .foregroundStyle(Color.ink)
                )
        }
        .buttonStyle(.plain)
    }
}

// MARK: - Testo lungo

/// Campo multiriga con contatore: nota della prenotazione e corpo della
/// campagna, che finora erano due box scritti a mano.
struct CounterTextField<Accessory: View>: View {
    let label: String
    @Binding var text: String
    var placeholder: String
    var limit: Int
    var error: String?
    var minHeight: CGFloat
    @ViewBuilder var accessory: Accessory

    init(
        label: String,
        text: Binding<String>,
        placeholder: String = "",
        limit: Int,
        error: String? = nil,
        minHeight: CGFloat = 56,
        @ViewBuilder accessory: () -> Accessory = { EmptyView() }
    ) {
        self.label = label
        self._text = text
        self.placeholder = placeholder
        self.limit = limit
        self.error = error
        self.minHeight = minHeight
        self.accessory = accessory()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                BrandSectionLabel(text: label)
                Spacer()
                Text(L("field_char_counter", text.count, limit))
                    .font(Typo.jost(11))
                    .foregroundStyle(text.count >= limit ? Color.errorRed : Color.textMuted)
            }
            VStack(alignment: .leading, spacing: 10) {
                TextField(placeholder, text: $text, axis: .vertical)
                    .font(Typo.jost(14))
                    .foregroundStyle(Color.ink)
                    .tint(Color.oliveWood)
                    .frame(minHeight: minHeight, alignment: .top)
                    .onChange(of: text) { _, newValue in
                        // Il limite è del backend: si taglia qui, non al salvataggio.
                        if newValue.count > limit { text = String(newValue.prefix(limit)) }
                    }
                accessory
            }
            .padding(14)
            .background(RoundedRectangle(cornerRadius: 16).fill(Color.bone))
            .overlay(
                RoundedRectangle(cornerRadius: 16)
                    .strokeBorder(error != nil ? Color.errorRed : Color.oliveWood, lineWidth: 1.5)
            )
            FieldMessageRow(error: error, helper: nil)
        }
    }
}

// MARK: - Nuovo cliente

/// Nome + cognome + telefono: la stessa scheda che aprono l'agenda
/// dell'operatore e la prenotazione manuale del titolare. Prima erano due copie
/// senza nessun controllo, e la creazione falliva in silenzio.
struct NewClientForm: View {
    let firstLabel: String
    let lastLabel: String
    let phoneLabel: String
    let ctaLabel: String
    @Binding var firstName: String
    @Binding var lastName: String
    @Binding var phone: String
    var errors: [String: String]
    var formError: String?
    var creating: Bool = false
    let onCreate: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(alignment: .top, spacing: 10) {
                NameField(label: firstLabel, text: $firstName, error: errors["firstName"])
                NameField(label: lastLabel, text: $lastName, error: errors["lastName"])
            }
            PhoneField(label: phoneLabel, text: $phone, error: errors["phone"])
            FormErrorBanner(message: formError)
            AccentButton(text: ctaLabel, action: onCreate, loading: creating, height: 48, corner: 14)
        }
    }
}

/// Le regole della scheda, in un posto solo: le usano entrambi i view model.
/// Il telefono è obbligatorio come su `POST /crm/clients`.
func validateNewClient(firstName: String, lastName: String, phone: String) -> [String: String] {
    var errors: [String: String] = [:]
    errors["firstName"] = validateName(firstName).message
    errors["lastName"] = validateName(lastName).message
    errors["phone"] = validatePhone(phone).message
    return errors.compactMapValues { $0 }
}

// MARK: - Errore di form

/// Banner di fallimento sotto un form: prima esisteva solo nelle schermate di
/// autenticazione, e le schermate del titolare non dicevano nulla.
struct FormErrorBanner: View {
    let message: String?

    var body: some View {
        if let message {
            HStack(spacing: 8) {
                Image(systemName: "exclamationmark.triangle")
                    .font(.system(size: 13))
                    .foregroundStyle(Color.errorRed)
                Text(message)
                    .font(Typo.bodySmall)
                    .foregroundStyle(Color.errorRed)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 12)
            .padding(.vertical, 10)
            .background(RoundedRectangle(cornerRadius: 12).fill(Color.errorRed.opacity(0.10)))
        }
    }
}

File: android/core/designsystem/src/main/kotlin/com/devora/mencare/core/designsystem/component/FormFields.kt
package com.devora.mencare.core.designsystem.component

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material.icons.outlined.Remove
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.devora.mencare.core.common.DURATION_STEP_MINUTES
import com.devora.mencare.core.common.PasswordStrength
import com.devora.mencare.core.common.ValidationError
import com.devora.mencare.core.common.formatDuration
import com.devora.mencare.core.common.formatPhone
import com.devora.mencare.core.common.normalizePhone
import com.devora.mencare.core.common.passwordStrength
import com.devora.mencare.core.common.sanitizePriceInput
import com.devora.mencare.core.designsystem.R
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.Stone

/**
 * L'unica famiglia di campi del prodotto: ogni form dell'app passa di qui, così
 * un concetto (telefono, email, password, prezzo, durata) si comporta e si
 * lamenta allo stesso modo ovunque. Le regole stanno in `:core:common`
 * (`Validation.kt`), la copy in `strings.xml` di questo modulo.
 */

/** Rende visibile la normalizzazione appena il campo perde il fuoco. */
@Composable
private fun Modifier.normalizeOnBlur(value: String, onNormalized: (String) -> Unit): Modifier {
    var hadFocus by remember { mutableStateOf(false) }
    return onFocusChanged { state ->
        if (hadFocus && !state.hasFocus) onNormalized(value)
        hadFocus = state.hasFocus
    }
}

/** Nome o cognome di una persona: iniziali maiuscole, trim all'uscita dal campo. */
@Composable
fun NameField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    error: ValidationError? = null,
    errorText: String? = null,
    placeholder: String? = null,
    enabled: Boolean = true,
    height: Dp = FieldHeight,
) {
    FilledTextField(
        value = value,
        onValueChange = onValueChange,
        label = label,
        modifier = modifier.normalizeOnBlur(value) { raw ->
            val trimmed = raw.trim()
            if (trimmed != raw) onValueChange(trimmed)
        },
        placeholder = placeholder,
        keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.Words),
        error = errorText ?: validationMessageOrNull(error),
        enabled = enabled,
        height = height,
    )
}

/** Email: tastiera dedicata, niente maiuscola automatica, trim + minuscole all'uscita. */
@Composable
fun EmailField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    error: ValidationError? = null,
    errorText: String? = null,
    placeholder: String? = null,
    leadingIcon: ImageVector? = null,
    enabled: Boolean = true,
    height: Dp = FieldHeight,
    trailing: @Composable (() -> Unit)? = null,
) {
    FilledTextField(
        value = value,
        onValueChange = onValueChange,
        label = label,
        modifier = modifier.normalizeOnBlur(value) { raw ->
            val normalized = raw.trim().lowercase()
            if (normalized != raw) onValueChange(normalized)
        },
        placeholder = placeholder,
        leadingIcon = leadingIcon,
        keyboardOptions = KeyboardOptions(
            keyboardType = KeyboardType.Email,
            capitalization = KeyboardCapitalization.None,
        ),
        error = errorText ?: validationMessageOrNull(error),
        enabled = enabled,
        height = height,
        trailing = trailing,
    )
}

/**
 * Telefono: tastiera numerica, prefisso `+39` finché il numero non ne porta uno
 * suo, e normalizzazione in E.164 quando il campo perde il fuoco — il valore
 * torna al chiamante già nella forma che il backend accetta.
 */
@Composable
fun PhoneField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    error: ValidationError? = null,
    errorText: String? = null,
    enabled: Boolean = true,
    height: Dp = FieldHeight,
    trailing: @Composable (() -> Unit)? = null,
) {
    val typed = value.trimStart()
    val hasOwnPrefix = typed.startsWith("+") || typed.startsWith("00")
    FilledTextField(
        value = value,
        onValueChange = onValueChange,
        label = label,
        modifier = modifier.normalizeOnBlur(value) { raw ->
            val pretty = normalizePhone(raw)?.let(::formatPhone)
            if (pretty != null && pretty != raw) onValueChange(pretty)
        },
        placeholder = stringResource(R.string.ds_phone_placeholder),
        prefix = if (hasOwnPrefix) null else stringResource(R.string.ds_phone_prefix),
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Phone),
        error = errorText ?: validationMessageOrNull(error),
        enabled = enabled,
        height = height,
        trailing = trailing,
    )
}

/**
 * Password: campo bordato con il toggle mostra/nascondi e, quando serve, il
 * misuratore di sicurezza calcolato dalla stessa funzione del backend.
 */
@Composable
fun PasswordField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    error: ValidationError? = null,
    errorText: String? = null,
    showStrength: Boolean = false,
    height: Dp = FieldHeight,
) {
    Column(modifier.fillMaxWidth()) {
        BorderedPasswordField(
            value = value,
            onValueChange = onValueChange,
            label = label,
            error = errorText ?: validationMessageOrNull(error),
            height = height,
        )
        if (showStrength && value.isNotEmpty()) {
            Spacer(Modifier.height(8.dp))
            PasswordStrengthMeter(passwordStrength(value))
        }
    }
}

@Composable
fun PasswordStrengthMeter(strength: PasswordStrength, modifier: Modifier = Modifier) {
    Row(modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        SegmentedMeter(
            filled = when (strength) {
                PasswordStrength.DEBOLE -> 1
                PasswordStrength.MEDIA -> 2
                PasswordStrength.FORTE -> 3
            },
            total = 4,
            modifier = Modifier.weight(1f),
        )
        Spacer(Modifier.width(7.dp))
        Text(
            passwordStrengthLabel(strength),
            style = MaterialTheme.typography.bodySmall.copy(fontSize = 11.sp),
            color = Ink,
        )
    }
}

/**
 * Prezzo in euro: tastiera decimale e filtro che lascia passare solo ciò che
 * può diventare un prezzo. La conversione in centesimi (arrotondata, mai
 * troncata) resta in `parsePriceEuros`.
 */
@Composable
fun PriceField(
    value: String,
    onValueChange: (String) -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    error: ValidationError? = null,
    errorText: String? = null,
    enabled: Boolean = true,
    height: Dp = FieldHeight,
) {
    FilledTextField(
        value = value,
        onValueChange = { onValueChange(sanitizePriceInput(it)) },
        label = label,
        modifier = modifier,
        prefix = stringResource(R.string.ds_price_prefix),
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
        error = errorText ?: validationMessageOrNull(error),
        enabled = enabled,
        height = height,
    )
}

/** Durata di un servizio: passi di 5 minuti, come la griglia degli slot. */
@Composable
fun DurationField(
    minutes: Int,
    onMinutesChange: (Int) -> Unit,
    modifier: Modifier = Modifier,
    error: ValidationError? = null,
    step: Int = DURATION_STEP_MINUTES,
    height: Dp = 56.dp,
) {
    Column(modifier.fillMaxWidth()) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(height)
                .clip(FieldShape)
                .background(Stone)
                .padding(horizontal = 14.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Text(
                formatDuration(minutes),
                style = MaterialTheme.typography.bodyLarge,
                color = Ink,
                modifier = Modifier.weight(1f),
            )
            StepperButton(
                icon = Icons.Outlined.Remove,
                contentDescription = stringResource(R.string.ds_duration_less),
                onClick = { onMinutesChange(minutes - step) },
            )
            StepperButton(
                icon = Icons.Outlined.Add,
                contentDescription = stringResource(R.string.ds_duration_more),
                onClick = { onMinutesChange(minutes + step) },
            )
        }
        FieldMessage(validationMessageOrNull(error))
    }
}

@Composable
private fun StepperButton(
    icon: ImageVector,
    contentDescription: String,
    onClick: () -> Unit,
) {
    Box(
        modifier = Modifier
            .size(28.dp)
            .clip(CircleShape)
            .background(Bone)
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Icon(icon, contentDescription = contentDescription, tint = Ink, modifier = Modifier.size(15.dp))
    }
}
